src/main/cpp/src/json_utils.cu

```diff
@@ -0,0 +1,236 @@
+      // replace it by a null. This is necessary for libcudf's JSON reader to work.
+      // Note that if we want to support ARRAY schema, we need to check for `[` instead.
+      auto constexpr start_character = '{';
+      if (not_eol && *itr != start_character) { return {false, true}; }
+
+      return {not_eol, not_eol};
+    });
+
+  auto constexpr num_levels  = 256;
+  auto constexpr lower_level = std::numeric_limits<char>::min();
+  auto constexpr upper_level = std::numeric_limits<char>::max();
+  auto const num_chars       = input.chars_size(stream);
+
+  rmm::device_uvector<uint32_t> d_histogram(num_levels, stream);
+  thrust::uninitialized_fill(
+    rmm::exec_policy_nosync(stream), d_histogram.begin(), d_histogram.end(), 0);
+
+  size_t temp_storage_bytes = 0;
+  cub::DeviceHistogram::HistogramEven(nullptr,
+                                      temp_storage_bytes,
+                                      input.chars_begin(stream),
+                                      d_histogram.begin(),
+                                      num_levels,
+                                      lower_level,
+                                      upper_level,
+                                      num_chars,
+                                      stream.value());
+  rmm::device_buffer d_temp(temp_storage_bytes, stream);
+  cub::DeviceHistogram::HistogramEven(d_temp.data(),
+                                      temp_storage_bytes,
+                                      input.chars_begin(stream),
+                                      d_histogram.begin(),
+                                      num_levels,
+                                      lower_level,
+                                      upper_level,
+                                      num_chars,
+                                      stream.value());
+
+  auto const zero_level = d_histogram.begin() - lower_level;
+
+  auto const find_first_zero_count_pos = [&](auto const begin, auto const end) {
+    return thrust::find(rmm::exec_policy_nosync(stream), begin, end, 0);
+  };
+
+  auto const find_first_non_existing_char_in_range = [&](auto const begin,
+                                                         auto const end) -> std::pair<char, bool> {
+    auto first_zero_count_pos = find_first_zero_count_pos(begin, end);
+    if (first_zero_count_pos == end) { return {'\0', false}; }
+
+    auto first_non_existing_char = static_cast<char>(first_zero_count_pos - zero_level);
+    while (!can_be_delimiter(first_non_existing_char)) {
+      first_zero_count_pos = find_first_zero_count_pos(first_zero_count_pos + 1, end);
+      if (first_zero_count_pos == end) { return {'\0', false}; }
+      first_non_existing_char = static_cast<char>(first_zero_count_pos - zero_level);
+    }
+    return {first_non_existing_char, true};
+  };
+
+  // Firstly, search from the `\n` character to the end of the histogram,
```

Comment:
nit: if the issues with not using \n as the delimiter are solved I would rather have us just start at 0 instead of trying really hard to use \n as the delimiter.